From e08a634f3fb16426895b4eae049fd6ee130f87e1 Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Mon, 25 Jan 2016 13:25:04 +0100
Subject: [PATCH] ss: add more TCP_INFO components

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1301148

commit 260804f422fd33aa78379270d564a495b7bb5717
Author: Eric Dumazet <edumazet@google.com>
Date:   Tue Jun 25 13:29:17 2013 -0700

    ss: add more TCP_INFO components

    Allow ss -i to display more TCP informations :

    unacked:N   Number of un-acked packets
    retrans:X/Y   X: number of outstanding retransmit packets
                  Y: total number of retransmits for the session
    lost:N       Number of lost packets (tcpi_lost)
    sacked:N     Number of sacked packets (tcpi_sacked)
    facked:N     Number of facked packets (tcpi_facked)
    reordering:N Reordering level (if different of 3)

    Example :

    $ ss -emoi dst 10.7.7.83
    tcp   ESTAB      0      1154056   10.7.7.84:54127    10.7.7.83:34342
    timer:(on,200ms,0) ino:57003 sk:ffff88063c51d0c0 <->
    	 skmem:(r0,rb89280,t0,tb2097152,f726504,w1436184,o0,bl0) ts sack cubic
    wscale:7,6 rto:310 rtt:107.375/1 mss:1448 cwnd:568 ssthresh:108 send
    61.3Mbps unacked:568 retrans:0/21 reordering:127 rcv_space:29200

    Signed-off-by: Eric Dumazet <edumazet@google.com>
    Cc: Yuchung Cheng <ycheng@google.com>
    Cc: Neal Cardwell <ncardwell@google.com>

Signed-off-by: Phil Sutter <psutter@redhat.com>
---
 misc/ss.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/misc/ss.c b/misc/ss.c
index 9aa0ada15e6e0..3a4083b3104c5 100644
--- a/misc/ss.c
+++ b/misc/ss.c
@@ -1351,6 +1351,19 @@ static void tcp_show_info(const struct nlmsghdr *nlh, struct inet_diag_msg *r)
 					 / rtt));
 		}
 
+		if (info->tcpi_unacked)
+			printf(" unacked:%u", info->tcpi_unacked);
+		if (info->tcpi_retrans || info->tcpi_total_retrans)
+			printf(" retrans:%u/%u", info->tcpi_retrans,
+			       info->tcpi_total_retrans);
+		if (info->tcpi_lost)
+			printf(" lost:%u", info->tcpi_lost);
+		if (info->tcpi_sacked && r->idiag_state != SS_LISTEN)
+			printf(" sacked:%u", info->tcpi_sacked);
+		if (info->tcpi_fackets)
+			printf(" fackets:%u", info->tcpi_fackets);
+		if (info->tcpi_reordering != 3)
+			printf(" reordering:%d", info->tcpi_reordering);
 		if (info->tcpi_rcv_rtt)
 			printf(" rcv_rtt:%g", (double) info->tcpi_rcv_rtt/1000);
 		if (info->tcpi_rcv_space)
-- 
2.6.4

